# MARATÓN BEHIND THE CODE 2021

# Desafío 03 - Recomendación de productos financieros | GFT

El desafío consiste en agregar las bases de estas instituciones que brinda el nuevo entorno Open Finance, procesar los datos y desarrollar modelos de recomendación de productos utilizando algoritmos como Apriori y AssociationRules, considerando una confianza en la regla del 80% y un soporte mínimo del 10% con un máximo de 5 antecedentes.

Las personas de la base estarán respaldadas por este modelo y se deberá someter a análisis el archivo de respuestas con las recomendaciones, con lo cual se verificará el porcentaje de respuestas correctas para cada recomendación.

Recuerda que muchas personas se quedarán sin una recomendación.
<hr>


## Exploración de dataset


In [1]:
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

!pip install mlxtend



In [2]:
# Cargue aqui el dataset
!wget --no-check-certificate --content-disposition https://raw.githubusercontent.com/maratonadev/desafio-3-2021/main/assets/data/InsuranceCompanyABC.csv
df_insurance = pd.read_csv(r'InsuranceCompanyABC.csv')

!wget --no-check-certificate --content-disposition https://raw.githubusercontent.com/maratonadev/desafio-3-2021/main/assets/data/InvestmentBankCDE.csv
df_investment = pd.read_csv(r'InvestmentBankCDE.csv')

!wget --no-check-certificate --content-disposition https://raw.githubusercontent.com/maratonadev/desafio-3-2021/main/assets/data/RetailBankEFG.csv
df_retail = pd.read_csv(r'RetailBankEFG.csv')


--2021-12-05 13:02:31--  https://raw.githubusercontent.com/maratonadev/desafio-3-2021/main/assets/data/InsuranceCompanyABC.csv
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 185.199.108.133, 185.199.111.133, 185.199.110.133, ...
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|185.199.108.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 541832 (529K) [text/plain]
Saving to: ‘InsuranceCompanyABC.csv’

InsuranceCompanyABC 100%[===================>] 529.13K  --.-KB/s    in 0.05s   

2021-12-05 13:02:31 (9.49 MB/s) - ‘InsuranceCompanyABC.csv’ saved [541832/541832]

--2021-12-05 13:02:31--  https://raw.githubusercontent.com/maratonadev/desafio-3-2021/main/assets/data/InvestmentBankCDE.csv
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 185.199.108.133, 185.199.109.133, 185.199.110.133, ...
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|185.199.108.133|:443... connected.
HTTP request s

Podemos usar los métodos .info () y .describe () para obtener información básica sobre la cantidad actual de datos, sus tipos y valores.

In [3]:
df_insurance.head()

ID   Idade     Renda  ... seguro vida Emp  seguro vida PF Seguro Residencial
0  71504.0  42.993  4428.512  ...               T               T                  T
1  59751.0  26.373  4489.930  ...               T               T                  T
2  11877.0  46.977  4567.741  ...               F               F                  F
3  33701.0  24.311  4592.132  ...               F               F                  F
4  62029.0  28.896  4723.197  ...               F               F                  F

[5 rows x 9 columns]

In [ ]:
df_insurance.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10082 entries, 0 to 10081
Data columns (total 9 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   ID                  10082 non-null  float64
 1   Idade               10082 non-null  float64
 2   Renda               10082 non-null  float64
 3   Regiao              10082 non-null  object 
 4   Genero              10082 non-null  float64
 5   seguro auto         10082 non-null  object 
 6   seguro vida Emp     10082 non-null  object 
 7   seguro vida PF      10082 non-null  object 
 8   Seguro Residencial  10082 non-null  object 
dtypes: float64(4), object(5)
memory usage: 709.0+ KB


In [ ]:
df_insurance.describe()

ID         Idade          Renda        Genero
count   10082.000000  10082.000000   10082.000000  10082.000000
mean    51041.608907     38.461186    7783.396381      0.548601
std     28587.314369     12.212422    7303.024782      0.497657
min      1004.000000     18.009000    2312.000000      0.000000
25%     26815.500000     28.925500    6971.313000      0.000000
50%     50765.500000     37.229500    7522.556000      1.000000
75%     75605.750000     46.639500    8061.152000      1.000000
max    100981.000000     85.605000  580000.000000      1.000000

In [ ]:
df_insurance.nunique()

ID                    9530
Idade                 8903
Renda                 9995
Regiao                   5
Genero                   2
seguro auto              2
seguro vida Emp          2
seguro vida PF           2
Seguro Residencial       2
dtype: int64

In [ ]:
df_investment.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10082 entries, 0 to 10081
Data columns (total 8 columns):
 #   Column                            Non-Null Count  Dtype  
---  ------                            --------------  -----  
 0   ID                                10082 non-null  float64
 1   Investimento Fundos_cambiais      10082 non-null  object 
 2   Investimento Fundos_commodities   10082 non-null  object 
 3   Investimento LCI                  10082 non-null  object 
 4   Investimento LCA                  10082 non-null  object 
 5   Investimento Poupanca             10082 non-null  object 
 6   Investimento Fundos Multimercado  10082 non-null  object 
 7   Investimento Tesouro Direto       10082 non-null  object 
dtypes: float64(1), object(7)
memory usage: 630.2+ KB


In [ ]:
df_investment.describe()

ID
count   10082.000000
mean    51041.608907
std     28587.314369
min      1004.000000
25%     26815.500000
50%     50765.500000
75%     75605.750000
max    100981.000000

In [ ]:
df_investment.nunique()

ID                                  9530
Investimento Fundos_cambiais           2
Investimento Fundos_commodities        2
Investimento LCI                       2
Investimento LCA                       2
Investimento Poupanca                  2
Investimento Fundos Multimercado       2
Investimento Tesouro Direto            2
dtype: int64

In [ ]:
df_retail.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10082 entries, 0 to 10081
Data columns (total 10 columns):
 #   Column                       Non-Null Count  Dtype  
---  ------                       --------------  -----  
 0   ID                           10082 non-null  float64
 1   Financiamento Casa           10082 non-null  object 
 2   Financiamento Carro          10082 non-null  object 
 3   Emprestimo _pessoal          10082 non-null  object 
 4   Emprestimo _consignado       10082 non-null  object 
 5   Emprestimo _limite_especial  10082 non-null  object 
 6   Emprestimo _educacao         10082 non-null  object 
 7   Emprestimo _viagem           10082 non-null  object 
 8   Investimento CDB             10082 non-null  object 
 9   Investimento Fundos          10082 non-null  object 
dtypes: float64(1), object(9)
memory usage: 787.8+ KB


In [ ]:
df_retail.describe()

ID
count   10082.000000
mean    51041.608907
std     28587.314369
min      1004.000000
25%     26815.500000
50%     50765.500000
75%     75605.750000
max    100981.000000

In [ ]:
df_retail.nunique()

ID                             9530
Financiamento Casa                2
Financiamento Carro               2
Emprestimo _pessoal               2
Emprestimo _consignado            2
Emprestimo _limite_especial       2
Emprestimo _educacao              2
Emprestimo _viagem                2
Investimento CDB                  2
Investimento Fundos               2
dtype: int64

Se observa que los 3 datasets comparten las siguientes caracteristicas:
 - Tienen la misma cantidad de registros, 10082
 - Misma cantidad de valores unicos para ID, 9530, es decir que tienen la misma cantidad de ID duplicados.
 - Los ID's estan en el mismo orden en los 3 datasets.

 Teniendo en cuenta esto, hacer un merge por ID no va a resultar, se van a generar mas datos y se van a mergear filas de una transacción con otra que no corresponde. Viendo esta particularidades, asumo cada fila se relaciona con la misma fila de los otros datasets, es decir juntar por index de la fila, todas las filas 1 de todos los datasets, las filas dos y asi, por las dudas verificar que luego los ID se correspondan, para validar que esta bien realizo el merge de los datasets.

In [4]:
df_merged = pd.concat([df_insurance, df_investment, df_retail], axis=1, sort=False)

In [5]:
df_merged.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10082 entries, 0 to 10081
Data columns (total 27 columns):
 #   Column                            Non-Null Count  Dtype  
---  ------                            --------------  -----  
 0   ID                                10082 non-null  float64
 1   Idade                             10082 non-null  float64
 2   Renda                             10082 non-null  float64
 3   Regiao                            10082 non-null  object 
 4   Genero                            10082 non-null  float64
 5   seguro auto                       10082 non-null  object 
 6   seguro vida Emp                   10082 non-null  object 
 7   seguro vida PF                    10082 non-null  object 
 8   Seguro Residencial                10082 non-null  object 
 9   ID                                10082 non-null  float64
 10  Investimento Fundos_cambiais      10082 non-null  object 
 11  Investimento Fundos_commodities   10082 non-null  object 
 12  Inve

In [6]:
df_merged.nunique()

ID                                  9530
Idade                               8903
Renda                               9995
Regiao                                 5
Genero                                 2
seguro auto                            2
seguro vida Emp                        2
seguro vida PF                         2
Seguro Residencial                     2
ID                                  9530
Investimento Fundos_cambiais           2
Investimento Fundos_commodities        2
Investimento LCI                       2
Investimento LCA                       2
Investimento Poupanca                  2
Investimento Fundos Multimercado       2
Investimento Tesouro Direto            2
ID                                  9530
Financiamento Casa                     2
Financiamento Carro                    2
Emprestimo _pessoal                    2
Emprestimo _consignado                 2
Emprestimo _limite_especial            2
Emprestimo _educacao                   2
Emprestimo _viag

In [ ]:
df_merged.to_csv(path_or_buf='merged.csv')

In [7]:
df_merged.drop(columns=['ID'],inplace=True)
df_merged.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10082 entries, 0 to 10081
Data columns (total 24 columns):
 #   Column                            Non-Null Count  Dtype  
---  ------                            --------------  -----  
 0   Idade                             10082 non-null  float64
 1   Renda                             10082 non-null  float64
 2   Regiao                            10082 non-null  object 
 3   Genero                            10082 non-null  float64
 4   seguro auto                       10082 non-null  object 
 5   seguro vida Emp                   10082 non-null  object 
 6   seguro vida PF                    10082 non-null  object 
 7   Seguro Residencial                10082 non-null  object 
 8   Investimento Fundos_cambiais      10082 non-null  object 
 9   Investimento Fundos_commodities   10082 non-null  object 
 10  Investimento LCI                  10082 non-null  object 
 11  Investimento LCA                  10082 non-null  object 
 12  Inve

### Analisis por tipo de variable


## Procesamiento de datos


## Generacion de frecuencia de items



In [58]:
from mlxtend.frequent_patterns import apriori

df_merged.replace(to_replace='T', value=True, inplace=True)
df_merged.replace(to_replace='F', value=False, inplace=True)

df_merged_temp = df_merged.drop(columns=['Idade', 'Renda', 'Regiao', 'Genero'])
frequent_itemsets = apriori(df_merged_temp, min_support=0.1, max_len=6, use_colnames=True)


df_merged_temp_hombres = df_merged[df_merged['Genero'] == 1].drop(columns=['Idade', 'Renda', 'Regiao', 'Genero'])
df_merged_temp_mujeres = df_merged[df_merged['Genero'] == 0].drop(columns=['Idade', 'Renda', 'Regiao', 'Genero'])

frequent_itemsets_hombres = apriori(df_merged_temp_hombres, min_support=0.1, max_len=6, use_colnames=True)

frequent_itemsets_mujeres = apriori(df_merged_temp_mujeres, min_support=0.1, max_len=6, use_colnames=True)

regiaos = ["CO", "SE", "NE", "S", "N"]
rules_by_region = {}
for region in regiaos:
  aux = df_merged.drop(columns=['Idade', 'Renda', 'Genero'])
  aux = aux[ aux["Regiao"] == region ]
  frequent_itemsets_aux = apriori(aux.drop(columns=['Regiao']), min_support=0.1, max_len=6, use_colnames=True)
  rules_by_region[region] = association_rules(frequent_itemsets_aux, metric="confidence", min_threshold=0.8)



In [16]:
from mlxtend.frequent_patterns import association_rules


rules = association_rules(frequent_itemsets, metric="confidence", min_threshold=0.8)

rules_hombres = association_rules(frequent_itemsets_hombres, metric="confidence", min_threshold=0.8)

rules_mujeres = association_rules(frequent_itemsets_mujeres, metric="confidence", min_threshold=0.8)

In [48]:
from collections import Counter

products = [
  "seguro auto",
  "seguro vida Emp",
  "seguro vida PF",
  "Seguro Residencial",
  "Investimento Fundos_cambiais",
  "Investimento Fundos_commodities",
  "Investimento LCI",
  "Investimento LCA",
  "Investimento Poupanca",
  "Investimento Fundos Multimercado",
  "Investimento Tesouro Direto",
  "Financiamento Casa",
  "Financiamento Carro",
  "Emprestimo _pessoal",
  "Emprestimo _consignado",
  "Emprestimo _limite_especial",
  "Emprestimo _educacao",
  "Emprestimo _viagem",
  "Investimento CDB",
  "Investimento Fundos"
]

def get_products_contratados(transaction):
  productos_contratados = [product for product in products if transaction[product]]
  return productos_contratados


def get_rules_for(productos_contratados, rules):
  rule_to_aply = []
  for index, rule in rules.iterrows():
    check =  all(item in productos_contratados for item in rule['antecedents'])
    #if Counter(rule['antecedents']) == Counter(productos_contratados):
    if check:
      rule_to_aply.append(rule)
  df = pd.DataFrame(rule_to_aply)
  if len(rule_to_aply) > 0:
    return df.sort_values(by=['confidence'], ascending=False).head()
  else:
    return df


def get_sugerencias(rules_to_aply, productos_contratados):
  sugerencias = {}
  for index, rule in rules_to_aply.iterrows():
    for sugerencia in rule['consequents']:
      if sugerencia not in sugerencias and sugerencia not in productos_contratados:
        sugerencias[sugerencia] = rule['confidence']
      if len(sugerencias) >= 3:
        break
    if len(sugerencias) >= 3:
      break

  return sugerencias

### Genero Respuestas

In [49]:
!wget --no-check-certificate --content-disposition https://raw.githubusercontent.com/maratonadev/desafio-3-2021/main/assets/data/ANSWERS.csv
df_answers = pd.read_csv(r'ANSWERS.csv')

df_answers.info()


--2021-12-05 15:25:58--  https://raw.githubusercontent.com/maratonadev/desafio-3-2021/main/assets/data/ANSWERS.csv
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 185.199.108.133, 185.199.109.133, 185.199.110.133, ...
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|185.199.108.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 1051067 (1.0M) [text/plain]
Saving to: ‘ANSWERS.csv.3’

ANSWERS.csv.3       100%[===================>]   1.00M  --.-KB/s    in 0.06s   

2021-12-05 15:25:59 (16.3 MB/s) - ‘ANSWERS.csv.3’ saved [1051067/1051067]

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 13730 entries, 0 to 13729
Data columns (total 31 columns):
 #   Column                            Non-Null Count  Dtype  
---  ------                            --------------  -----  
 0   ID                                13730 non-null  float64
 1   Idade                             13730 non-null  float64
 2   Renda                      

In [67]:
df_answers_temp = df_answers.replace(to_replace='T', value=True, inplace=False)
df_answers_temp = df_answers_temp.replace(to_replace='F', value=False, inplace=False)

df_answers_temp['RECOMMENDATION_1'] = df_answers_temp['RECOMMENDATION_1'].astype(str)
df_answers_temp['RECOMMENDATION_2'] = df_answers_temp['RECOMMENDATION_2'].astype(str)
df_answers_temp['RECOMMENDATION_3'] = df_answers_temp['RECOMMENDATION_3'].astype(str)
df_answers_temp['CONFIDENCE_1'] = df_answers_temp['CONFIDENCE_1'].astype(str)
df_answers_temp['CONFIDENCE_2'] = df_answers_temp['CONFIDENCE_2'].astype(str)
df_answers_temp['CONFIDENCE_3'] = df_answers_temp['CONFIDENCE_3'].astype(str)

df_answers_temp['RECOMMENDATION_1'] = ""
df_answers_temp['RECOMMENDATION_2'] = ""
df_answers_temp['RECOMMENDATION_3'] = ""
df_answers_temp['CONFIDENCE_1'] = ""
df_answers_temp['CONFIDENCE_2'] = ""
df_answers_temp['CONFIDENCE_3'] = ""

conseguidas = 0
for index, transaction in df_answers_temp.iterrows():
  productos_contratados = get_products_contratados(transaction)
  
  rules_to_aply = get_rules_for(productos_contratados, rules)
  sugerencias = get_sugerencias(rules_to_aply, productos_contratados)
  if len(sugerencias) <= 0:
    rules_genero = rules_hombres if transaction['Genero'] == 1 else rules_mujeres
    rules_to_aply = get_rules_for(productos_contratados, rules_genero)
    sugerencias = get_sugerencias(rules_to_aply, productos_contratados)
  if len(sugerencias) <= 0:
    rules_to_aply = get_rules_for(productos_contratados, rules_by_region[transaction['Regiao']])
    sugerencias = get_sugerencias(rules_to_aply, productos_contratados)
  
  
  print(productos_contratados)
  for index_, rule in rules_to_aply.iterrows():
    print(rule['antecedents'])
  
  if len(sugerencias) > 0:
    print(sugerencias)
    recomendacion_numero = 1
    for sugerencia in sugerencias:
      df_answers_temp.at[index, 'RECOMMENDATION_{}'.format(recomendacion_numero)] = sugerencia
      df_answers_temp.at[index, 'CONFIDENCE_{}'.format(recomendacion_numero)] = sugerencias[sugerencia]
      recomendacion_numero = recomendacion_numero + 1


Streaming output truncated to the last 5000 lines.
frozenset({'Investimento LCI', 'Investimento Fundos_commodities'})
frozenset({'Investimento LCI', 'Investimento Fundos_commodities', 'Investimento Fundos'})
frozenset({'Investimento LCI', 'Investimento Fundos'})
{'Investimento Fundos_cambiais': 0.9379671150971598, 'Investimento CDB': 0.897608370702541}
['Investimento Fundos_cambiais', 'Investimento Fundos_commodities', 'Investimento CDB']
frozenset({'Investimento CDB', 'Investimento Fundos_commodities', 'Investimento Fundos_cambiais'})
frozenset({'Investimento Fundos_commodities', 'Investimento Fundos_cambiais'})
{'Investimento LCI': 0.9188118811881189}
['Investimento Fundos_cambiais', 'Investimento Fundos_commodities', 'Investimento CDB']
frozenset({'Investimento CDB', 'Investimento Fundos_commodities', 'Investimento Fundos_cambiais'})
frozenset({'Investimento Fundos_commodities', 'Investimento Fundos_cambiais'})
{'Investimento LCI': 0.9188118811881189}
['Investimento Fundos_commoditi

In [68]:
df_answers['RECOMMENDATION_1'] = df_answers_temp['RECOMMENDATION_1']
df_answers['RECOMMENDATION_2'] = df_answers_temp['RECOMMENDATION_2']
df_answers['RECOMMENDATION_3'] = df_answers_temp['RECOMMENDATION_3']
df_answers['CONFIDENCE_1'] = df_answers_temp['CONFIDENCE_1']
df_answers['CONFIDENCE_2'] = df_answers_temp['CONFIDENCE_2']
df_answers['CONFIDENCE_3'] = df_answers_temp['CONFIDENCE_3']

In [69]:
df_answers.to_csv(path_or_buf='answers_solucion.csv', index=False)

In [70]:
df_answers.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 13730 entries, 0 to 13729
Data columns (total 31 columns):
 #   Column                            Non-Null Count  Dtype  
---  ------                            --------------  -----  
 0   ID                                13730 non-null  float64
 1   Idade                             13730 non-null  float64
 2   Renda                             13730 non-null  float64
 3   Regiao                            13730 non-null  object 
 4   Genero                            13730 non-null  float64
 5   seguro auto                       13730 non-null  object 
 6   seguro vida Emp                   13730 non-null  object 
 7   seguro vida PF                    13730 non-null  object 
 8   Seguro Residencial                13730 non-null  object 
 9   Investimento Fundos_cambiais      13730 non-null  object 
 10  Investimento Fundos_commodities   13730 non-null  object 
 11  Investimento LCI                  13730 non-null  object 
 12  Inve

In [71]:
df_answers_temp.head()

ID   Idade     Renda  ... CONFIDENCE_2  RECOMMENDATION_3  CONFIDENCE_3
0  1004.0  40.685  8474.687  ...     0.876515                                
1  1028.0  56.693  7996.353  ...                                             
2  1034.0  49.413  8624.335  ...                                             
3  1035.0  22.764  6885.196  ...                                             
4  1039.0  22.059  7749.965  ...                                             

[5 rows x 31 columns]